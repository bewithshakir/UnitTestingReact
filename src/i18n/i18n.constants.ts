export const namespaces = {
    pages: { 
      simple : "pages.simple"
    }
  };
  
  export const languages = {
    es: "es",
    en: "en",
    fr : "fr"
  };