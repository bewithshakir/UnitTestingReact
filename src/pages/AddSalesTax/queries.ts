import { useMutation, useQuery } from "react-query";
import { AxiosRequestConfig } from "axios";
import axios from "../../infrastructure/ApiHelper";


const addSalesTax = async (payload: any) => {
    const options: AxiosRequestConfig = {
        method: 'post',
        url: 'api/tax-service/sales-taxes',
        data: payload,
    };
    const { data } = await axios(options);
    return data;
};


export const useAddSalesTax = (onError: any, onSuccess: any) => {
    return useMutation((payload: any) =>
        addSalesTax(payload), {
        onError,
        onSuccess,
        retry: false,
    });
};

// interface PayloadProps {

// }
const fetchSaleTax = async (query: any) => {
    const options: AxiosRequestConfig = {
        method: 'get',
        url: `api/tax-service/sales-taxes/cities${query}`
    };
    return axios(options);
};

export const useGetSaleTax = (query: any, onSuccess: any, onError: any) => {
    return useQuery(['get-sale-tax'], () => fetchSaleTax(query), {
        onSuccess,
        onError,
        enabled: !!query,
        retry: false,
    });
};

const editSaleTax = async (payload: any,) => {
    const options: AxiosRequestConfig = {
        method: 'put',
        url: `/api/tax-service/sales-taxes`,
        data: payload
    };
    const { data } = await axios(options);
    return data;
};

export const useEditSalesTax = (onSuccess: any, onError: any) => {
    return useMutation((payload: any) => editSaleTax(payload), {
        onSuccess,
        onError,
        retry: false
    });
};