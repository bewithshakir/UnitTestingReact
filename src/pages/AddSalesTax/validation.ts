import * as Yup from 'yup';

export const AddSalesTaxValidationSchema = Yup.object().shape({
    addressLine1: Yup.string().required('Required'),
    state: Yup.string().required('Required'),
    city: Yup.string().required('Required'),
    stateRate: Yup.string().matches(/^[0-9]+([.][0-9]+)?$/, 'Invalid Rate.').required('Required'),
    localRate: Yup.string().matches(/^[0-9]+([.][0-9]+)?$/, 'Invalid Rate.').required('Required')
});


export const AddSalesTaxValidationSchemaEdit = Yup.object().shape({
    addressLine1: Yup.string(),
    state: Yup.string().required('Required'),
    city: Yup.string().required('Required'),
    stateRate: Yup.string().matches(/^[0-9]+([.][0-9]+)?$/, 'Invalid Rate.').required('Required'),
    localRate: Yup.string().matches(/^[0-9]+([.][0-9]+)?$/, 'Invalid Rate.').required('Required')
});
