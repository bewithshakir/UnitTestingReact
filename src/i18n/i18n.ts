import i18next, { i18n as i18nInstance } from "i18next";
import { initReactI18next } from "react-i18next";
import { languages } from "./i18n.constants";
import HttpApi from "i18next-xhr-backend";
import LanguageDetector from 'i18next-browser-languagedetector';

const createI18n = (language: string): i18nInstance => {
  const i18n = i18next.createInstance().use(initReactI18next);

  i18n
    .use(HttpApi)
    .use(LanguageDetector)
    .init({
      // backend: {
      //   loadPath: "./locales/{{lng}}/{{ns}}.json", // Specify where backend will find translation files.
      // },
      lng: language,
      debug: true,
      fallbackLng: language,
      interpolation: {
        escapeValue: false,
      },
      // ns: namespaces.pages.simple,
    });

  return i18n;
};

export const i18n = createI18n(languages.en);
