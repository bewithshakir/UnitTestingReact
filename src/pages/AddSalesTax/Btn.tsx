import React from 'react';
import { useHistory } from 'react-router';
import { useAddSalesTax } from './queries';

const Btn: React.FC = ()=> {
    const history = useHistory();
    
    const onAddSalesTaxError = ()=> {
        // return
    };
    const onAddSalesTaxSuccess = ()=> {
        // return
    };
    const { mutate: addNewSalesTax } = useAddSalesTax(onAddSalesTaxError, onAddSalesTaxSuccess);
    const goToBtn = () => {
        const apiPayload = {
            "countryCode":"us",
            "state":"TX",
            "city":"Houstan",
            "stateRate":1.32,
            "federalRate":3.654,
            "localRate":2.3456
        };
        addNewSalesTax(apiPayload);
        history.push('/btn');
    };
    return (
        <div data-test="component-btn">
            Button from component
            <button type="button" onClick={goToBtn}>btn</button>
        </div>
    );
};
export default Btn;