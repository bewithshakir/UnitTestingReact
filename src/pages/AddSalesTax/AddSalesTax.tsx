import { Box, Container, Grid, Typography } from '@mui/material';
import { useFormik } from 'formik';
import React, { useState, useEffect } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import './style.scss';

import SalesTaxModel from '../../models/SalesTaxModel';
import AutocompleteInput from '../../components/UIComponents/GoogleAddressComponent/GoogleAutoCompleteAddress';
import Input from '../../components/UIComponents/Input/Input';
import { Button } from '../../components/UIComponents/Button/Button.component';
import {useAddSalesTax, useEditSalesTax, useGetSaleTax} from './queries';
import ToastMessage from '../../components/UIComponents/ToastMessage/ToastMessage.component';
import { HorizontalBarVersionState, useStore, useShowConfirmationDialogBoxStore } from '../../store';
import {AddSalesTaxValidationSchema, AddSalesTaxValidationSchemaEdit} from './validation';


const initialValues = new SalesTaxModel();

export interface AddSalesTaxProps{
    version:string
  }
  
interface IFormStatus {
    message: string
    type: string
}
interface IFormStatusProps {
    [key: string]: IFormStatus
}


const formStatusProps: IFormStatusProps = {
    editsuccess: {
        message: 'Data updated successfully',
        type: 'Success',
    },
    success: {
        message: 'Data added successfully.',
        type: 'Success',
    },
    updated: {
        message: 'Data updated successfully.',
        type: 'Success',
    },
    duplicate: {
        message: 'Customer Id already exist. Please use different Customer Id.',
        type: 'Error',
    },
    error: {
        message: 'Something went wrong. Please try again.',
        type: 'Error',
    }
};

const AddSalesTax: React.FC<AddSalesTaxProps> = () => {
    const setVersion = useStore((state: HorizontalBarVersionState) => state.setVersion);
    const isFormValidated = useShowConfirmationDialogBoxStore((state) => state.setFormFieldValue);
    const resetFormFieldValue = useShowConfirmationDialogBoxStore((state) => state.resetFormFieldValue);

    const isFormFieldChange = () => formik.dirty;
    setVersion("Breadcrumbs-Single");
    const [apiResposneState, setAPIResponse] = useState(false);
    // const [isDisabled, setDisabled] = useState(false);
    
    const navigate = useNavigate();
    const location = useLocation();
    const { t } = useTranslation();

    const [formStatus, setFormStatus] = useState<IFormStatus>({
        message: '',
        type: '',
    });

    useEffect(()=>{
        resetFormFieldValue();
    },[]);
    

    // edit section
    const [isEditMode, setEditMode] = useState(false);
    

    const populateDataInAllFields = (formData: any)=> {
        formik.setFieldValue('addressLine1', formData.addressLine1);
        formik.setFieldValue('city', formData.city);
        formik.setFieldValue('state', formData.state);
        formik.setFieldValue('countryCd', 'us');
        formik.setFieldValue('federalRate', formData.federalRate);
        formik.setFieldValue('stateRate', formData.stateRate);
        formik.setFieldValue('localRate', formData.localRate);
    };

    const onGetSaleTaxSuccess = (response: any)=> {
        populateDataInAllFields(response.data.data);
        setEditMode(true);
    };
    const onGetSaleTaxError = (err: any)=> {
        try {
            const { data } = err.response;
            setAPIResponse(true);
            setFormStatus({ message: data?.error?.message || formStatusProps.error.message, type: 'Error' });
            formik.setSubmitting(false);
            setTimeout(() => {
                setAPIResponse(false);
            }, 6000);
        } catch(error) {
            setFormStatus(formStatusProps.error);
        }

    };
    
    useGetSaleTax(location.search, onGetSaleTaxSuccess, onGetSaleTaxError);
    

    const onEditSaleTaxSuccess = ()=> {
        setAPIResponse(true);
        isFormValidated(false);
        setFormStatus(formStatusProps.updated);
        setTimeout(() => {
            setAPIResponse(false);
        }, 6000);
    };
    const onEditSaleTaxError = (err: any)=> {
        try {
            const { data } = err.response;
            setAPIResponse(true);
            // setDisabled(false);
            setFormStatus({ message: data?.error?.message || formStatusProps.error.message, type: 'Error' });
            formik.setSubmitting(false);
            setTimeout(() => {
                setAPIResponse(false);
            }, 6000);
        } catch (error) {
            setFormStatus(formStatusProps.error);
        }
    };

    const { mutate: editSaleTax } = useEditSalesTax(onEditSaleTaxSuccess, onEditSaleTaxError);

    const updateSaleTax = (form: SalesTaxModel)=> {
        try {
            const payload = {
                "countryCode": form.countryCd,
                "city": form.city,
                "state": form.state,
                "stateRate": form.federalRate ? parseFloat(form.stateRate): 0,
                "federalRate": form.federalRate ? parseFloat(form.federalRate) : 0,
                "localRate": form.federalRate ? parseFloat(form.localRate): 0
            };
            editSaleTax(payload);
        } catch (error) {
            setFormStatus(formStatusProps.error);
        }
    };

    // Edit end


    const onAddSalesTaxSuccess = () => {
        setAPIResponse(true);
        isFormValidated(false);
        // setDisabled(true);
        setFormStatus(formStatusProps.success);
        // formik.resetForm({});
        setTimeout(() => {
            setAPIResponse(false);
        }, 6000);
    };
    const onAddSalesTaxError = (err: any) => {
        try {
            const { data } = err.response;
            setAPIResponse(true);
            // setDisabled(false);
            setFormStatus({ message: data?.error?.message || formStatusProps.error.message, type: 'Error' });
            formik.setSubmitting(false);
            setTimeout(() => {
                setAPIResponse(false);
            }, 6000);
        } catch (error) {
            setFormStatus(formStatusProps.error);
        }
    };

    const { mutate: addNewSalesTax } = useAddSalesTax(onAddSalesTaxError, onAddSalesTaxSuccess);

    const createNewSalesTax = (form: SalesTaxModel) => {
        try {
            const apiPayload = {
                "countryCode": form.countryCd,
                "city": form.city,
                "state": form.state,
                "stateRate": parseFloat(form.stateRate),
                "federalRate": form.federalRate ? parseFloat(form.federalRate): 0,
                "localRate": parseFloat(form.localRate)
            };
            addNewSalesTax(apiPayload);
            
        } catch (error) {
            setFormStatus(formStatusProps.error);
        }
    };

    const formik = useFormik({
        initialValues,
        validationSchema: isEditMode ? AddSalesTaxValidationSchemaEdit: AddSalesTaxValidationSchema,
        onSubmit: (values) => {
            if(isEditMode) {
                updateSaleTax(values);
            } else {
                createNewSalesTax(values);
            }
        }
    });
    
    const handleFormDataChange = () => {
        if (isEditMode) {
                if (formik.dirty) {
                    if (formik.initialValues != formik.values) {
                        isFormValidated(false);
                    }
                }
        } 
        if (isFormFieldChange() && !formik.isSubmitting) {
            isFormValidated(true); 
        }
    };    

    const handleGoogleAddressChange = (addressObj: any) => {
        formik.setFieldValue('addressLine1', addressObj.addressLine1);
        formik.setFieldValue('city', addressObj.city);
        formik.setFieldValue('state', addressObj.state);
        formik.setFieldValue('countryCd', 'us');
    };
    const handleGoogleAddressBlur = () => {
        formik.setFieldTouched("addressLine1");
        formik.validateField("addressLine1");

        formik.setFieldTouched("city");
        formik.validateField("city");

        formik.setFieldTouched("state");
        formik.validateField("state");

        formik.setFieldTouched("stateRate");
        formik.validateField("stateRate");

        formik.validateField("federalRate");
        formik.validateField("federalRate");

        formik.validateField("localRate");
        formik.validateField("localRate");
    };
    

    function onClickBack () {
        navigate('/salesTax');
    }
    const disableButton = () => {
        if (isEditMode) {
            if (Object.keys(formik.errors).length > 1) {
                return true;
            } 
            else if ((formik.touched.stateRate && formik.errors.stateRate) ||
            (formik.touched.federalRate && formik.errors.federalRate) || 
            (formik.touched.localRate && formik.errors.localRate)) {
                return true;
            }
            else if (Object.keys(formik.touched).length) {
                return false;
            }
        } else {
            if ((formik.touched.stateRate && formik.errors.stateRate) ||
            (formik.touched.federalRate && formik.errors.federalRate) ||
            (formik.touched.localRate && formik.errors.localRate)) {
                return true;
            }
            else if (formik.values.addressLine1 && 
                formik.values.city && 
                formik.values.state && 
                formik.values.stateRate &&
                formik.values.federalRate &&
                formik.values.localRate
                ){
                return false;
            } else {
                return true;
            }
        }
    };

    return (
        <Box display="flex" className="global_main_wrapper">
            <Grid item md={10} xs={10}>
                <Container maxWidth="lg" className="page-container">
                   
                    <form onSubmit={formik.handleSubmit} onBlur={handleFormDataChange} data-test="component-AddSalesTax" >
                        <Typography color="var(--Darkgray)" variant="h3" gutterBottom className="fw-bold" mb={1} pt={3}>
                        {t("taxes.salesTax.form.title")} *
                        </Typography>
                        <Grid container mt={1}>
                            <Grid item xs={12} md={12} pr={2.5} pb={2.5}>
                                <Grid item xs={12} md={6}>
                                    <AutocompleteInput
                                        name='addressLine1'
                                        label={t("taxes.salesTax.form.labelLocation")}
                                        onChange={handleGoogleAddressChange}
                                        onBlur={handleGoogleAddressBlur}
                                        value={formik.values.addressLine1}
                                        helperText={(formik.touched.addressLine1 && formik.errors.addressLine1) ? formik.errors.addressLine1 : undefined}
                                        error={(formik.touched.addressLine1 && formik.errors.addressLine1) ? true : false}
                                        disabled={isEditMode}
                                        data-test="auto-complete-input"
                                    />
                                </Grid>
                            </Grid>
                            <Grid item xs={12} md={12} pr={2.5} pb={2.5}>
                                <Grid item xs={12} md={6}>
                                    <Input
                                        id='city'
                                        label={t("taxes.salesTax.form.labelCity")}
                                        type='text'
                                        disabled
                                        helperText={(formik.touched.city && formik.errors.city) ? formik.errors.city : undefined}
                                        error={(formik.touched.city && formik.errors.city) ? true : false}
                                        description=''
                                        {...formik.getFieldProps('city')}
                                        data-test="city"
                                        required
                                    />

                                </Grid>
                            </Grid>
                            <Grid item xs={12} md={12} pr={2.5} pb={2.5}>
                                <Grid item xs={12} md={6}>
                                    <Input
                                        id='state'
                                        label={t("taxes.salesTax.form.labelState")}
                                        type='text'
                                        description=''
                                        disabled
                                        helperText={(formik.touched.state && formik.errors.state) ? formik.errors.state : undefined}
                                        error={(formik.touched.state && formik.errors.state) ? true : false}
                                        {...formik.getFieldProps('state')}
                                        data-test="state"
                                        required
                                    />
                                </Grid>
                            </Grid>

                            <Grid item xs={12} md={12} pr={2.5} pb={2.5}>
                                <Grid item xs={12} md={6}>
                                    <Input
                                        id='stateRate'
                                        label={t("taxes.salesTax.form.labelStateRate")+'%'}
                                        type='text'
                                        placeholder='State Rate'
                                        helperText={(formik.touched.stateRate && formik.errors.stateRate) ? formik.errors.stateRate : undefined}
                                        error={(formik.touched.stateRate && formik.errors.stateRate) ? true : false}
                                        {...formik.getFieldProps('stateRate')}
                                        required
                                        data-test="stateRate"
                                    />
                                </Grid>
                            </Grid>

                            <Grid item xs={12} md={12} pr={2.5} pb={2.5}>
                                <Grid item xs={12} md={6}>
                                    <Input
                                        id='federalRate'
                                        label={t("taxes.salesTax.form.labelFederalRate")+'%'}
                                        type='text'
                                        placeholder='Federal Rate'
                                        helperText={(formik.touched.federalRate && formik.errors.federalRate) ? formik.errors.federalRate : undefined}
                                        error={(formik.touched.federalRate && formik.errors.federalRate) ? true : false}
                                        {...formik.getFieldProps('federalRate')}
                                        required
                                        data-test="federalRate"
                                    />
                                </Grid>
                            </Grid>

                            <Grid item xs={12} md={12} pr={2.5} pb={2.5}>
                                <Grid item xs={12} md={6}>
                                    <Input
                                        id='localRate'
                                        label={t("taxes.salesTax.form.labelLocalRate")+'%'}
                                        type='text'
                                        placeholder='Local Rate'
                                        helperText={(formik.touched.localRate && formik.errors.localRate) ? formik.errors.localRate : undefined}
                                        error={(formik.touched.localRate && formik.errors.localRate) ? true : false}
                                        {...formik.getFieldProps('localRate')}
                                        required
                                        data-test="localRate"
                                    />
                                </Grid>
                            </Grid>
                        </Grid>
                        <Grid container mt={1} justifyContent="flex-end">
                            <Grid item xs={12} md={6} pr={2.5} pb={2.5}>
                                <Box className="form-action-section">
                                    <Button
                                        types="cancel"
                                        aria-label={t("buttons.cancel")}
                                        className="mr-4"
                                        onClick={onClickBack}
                                        data-test="cancel"
                                    >
                                        {t("buttons.cancel")}
                                    </Button>
                                    <Button
                                        type="submit"
                                        types="save"
                                        aria-label={t("buttons.save")}
                                        className="ml-4"
                                        data-test="save"
                                        disabled={disableButton()}
                                    >
                                        {t("buttons.save")}
                                    </Button>
                                    
                                </Box>
                                <ToastMessage isOpen={apiResposneState} messageType={formStatus.type} onClose={() => { return ''; }} message={formStatus.message} />
                            </Grid>
                        </Grid>
                    </form>
                </Container>
            </Grid>
        </Box>
    );
};
export default AddSalesTax;