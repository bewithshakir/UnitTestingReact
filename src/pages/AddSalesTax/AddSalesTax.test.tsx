import { mount, shallow } from 'enzyme';
import { I18nextProvider } from 'react-i18next';
import { QueryClient, QueryClientProvider, useQuery } from 'react-query';

import {i18n} from '../../i18n/i18n';
import { findByIdAttr, findByTestAttr } from '../../tests/testUtils';
import AddSalesTax from './AddSalesTax';

const queryClient = new QueryClient();

/**
 * Factory function to create wrapper
 * @returns {wrapper}
 */

const setup = (props={})=>  mount(<QueryClientProvider client={queryClient}>
    <I18nextProvider i18n={i18n}>
        <AddSalesTax {...props}/>
    </I18nextProvider>
</QueryClientProvider>);


describe('AddSalesTax component', () => {

    it('renders component without error ', () => {
        const wrapper = setup();
        const component = findByTestAttr(wrapper, 'component-AddSalesTax');
        expect(component.length).toBe(1);
    });

    it('renders auto input without error', ()=>{
        const wrapper = setup(); 
        const autoInput = findByTestAttr(wrapper, 'auto-complete-input');
        expect(autoInput.length).toBe(1);
    });

    it('renders city input without error', ()=>{
        const wrapper = setup(); 
        const cityInput = findByTestAttr(wrapper, 'city');
        expect(cityInput.length).toBe(1);
    });

    it('renders state input without error', ()=>{
        const wrapper = setup(); 
        const cityInput = findByTestAttr(wrapper, 'state');
        expect(cityInput.length).toBe(1);
    });

    it('renders stateRate input without error', ()=>{
        const wrapper = setup(); 
        const cityInput = findByTestAttr(wrapper, 'stateRate');
        expect(cityInput.length).toBe(1);
    });

    it('renders localRate input without error', ()=>{
        const wrapper = setup(); 
        const cityInput = findByTestAttr(wrapper, 'localRate');
        expect(cityInput.length).toBe(1);
    });

    it('renders Cancel Button without error', ()=>{
        const wrapper = setup(); 
        const cancelButton = findByTestAttr(wrapper, 'cancel');
        expect(cancelButton.length).toBeGreaterThanOrEqual(0);
    });

    it('renders Save Button without error', ()=>{
        const wrapper = setup(); 
        const cancelButton = findByTestAttr(wrapper, 'save');
        expect(cancelButton.length).toBeGreaterThanOrEqual(0);
    });

    

    /* describe('renders city with all cases', ()=> {
        it('renders city input without error', ()=>{
            const wrapper = setup(); 
            const cityInput = findByTestAttr(wrapper, 'city');
            expect(cityInput.length).toBe(1);
        });
    }); */

    // it('renders city input with `disabled` attributes', ()=>{
    //     const wrapper = setup(); 
    //     const cityInput = findByIdAttr(wrapper, 'city');
    //     expect(cityInput.closest('input')).toHaveAttribute('disabled');
    //     // console.log('city--', cityInput.closest('input'))
    // });
    

});