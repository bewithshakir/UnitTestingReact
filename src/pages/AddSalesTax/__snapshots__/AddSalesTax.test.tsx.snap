// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<ForwardRef(Dialog)
  BackdropProps={
    Object {
      "style": Object {
        "backgroundColor": "gray",
        "opacity": 0.3,
      },
    }
  }
  aria-describedby="discard-changes-dialog-description"
  aria-labelledby="discard-changes-dialog-title"
  open={true}
>
  <div
    className="discard-dialog-container"
  >
    <ForwardRef(Typography)
      className="fw-bold"
      color="var(--Darkgray)"
      component="h2"
      pt={2.5}
      px={2.5}
      variant="h2"
    >
      Discard dialog
    </ForwardRef(Typography)>
    <ForwardRef(DialogContent)
      className="dialog-content"
    >
      <ForwardRef(DialogContentText)
        color="var(--Darkgray)"
        id="discard-changes-dialog-description"
      >
        content
      </ForwardRef(DialogContentText)>
    </ForwardRef(DialogContent)>
    <ForwardRef(DialogActions)
      className="dialog-actions"
    >
      <Button
        className="action-no"
        onClick={[Function]}
        variant="outlined"
      >
        Cancel
      </Button>
      <Button
        className="action-yes"
        color="default"
        onClick={[Function]}
      >
        Discard
      </Button>
    </ForwardRef(DialogActions)>
  </div>
</ForwardRef(Dialog)>
`;
